feat: add `ml/strided/dkmeans-distance` (#13062)
stdlib-js/stdlib: lib/node_modules/@stdlib/ml/strided/dkmeans-distance/README.md

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dkmeansDistance

> > Computes the distance between two double-precision floating-point strided arrays given a kmeans supported distance metric.

Reviewer: ```suggestion
> Computes the distance between two double-precision floating-point strided arrays according to a specified distance metric.
```

<section class="intro">

> 

Reviewer: ```suggestion
```

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var dkmeansDistance = require( '@stdlib/ml/strided/dkmeans-distance' );
```

#### dkmeansDistance( N, metric, x, strideX, y, strideY )

> Computes the distance between two double-precision floating-point strided arrays given a kmeans supported distance metric.

Reviewer: ```suggestion
Computes the distance between two double-precision floating-point strided arrays according to a specified distance metric.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
var y = new Float64Array( [ 2.0, 1.0, 2.0, 1.0, -2.0, 2.0, 3.0, 4.0 ] );

var z = dkmeansDistance( x.length, 'sqeuclidean', x, 1, y, 1 );
// returns 72.0
```

The function has the following parameters:

-   **N**: number of indexed elements.

> -   **metric**: distance metric.

Reviewer: ```suggestion
-   **metric**: distance [metric][@stdlib/ml/base/kmeans/metrics].
```

-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length of `x`.
-   **y**: input [`Float64Array`][@stdlib/array/float64].
-   **strideY**: stride length of `y`.

> The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to calculate the squared Euclidean distance between every other element in `x` and the first `N` elements of `y` in reverse order,

Reviewer: ```suggestion
The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to calculate the distance between every other element in `x` and the first `N` elements of `y` in reverse order,
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
var y = new Float64Array( [ 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 ] );

var z = dkmeansDistance( 3, 'sqeuclidean', x, 2, y, -1 );
// returns 20.0
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial arrays...
var x0 = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
var y0 = new Float64Array( [ 7.0, 8.0, 9.0, 10.0, 11.0, 12.0 ] );

// Create offset views...
var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var y1 = new Float64Array( y0.buffer, y0.BYTES_PER_ELEMENT*3 ); // start at 4th element

var z = dkmeansDistance( 3, 'sqeuclidean', x1, 1, y1, 1 );
// returns 192.0
```

#### dkmeansDistance.ndarray( N, metric, x, strideX, offsetX, y, strideY, offsetY )

> Computes the distance between two double-precision floating-point strided arrays given a kmeans supported distance metric using alternative indexing semantics.

Reviewer: ```suggestion
Computes the distance between two double-precision floating-point strided arrays according to a specified distance metric using alternative indexing semantics.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
var y = new Float64Array( [ 2.0, 1.0, 2.0, 1.0, -2.0, 2.0, 3.0, 4.0 ] );

var z = dkmeansDistance.ndarray( x.length, 'sqeuclidean', x, 1, 0, y, 1, 0 );
// returns 72.0
```

The function has the following additional parameters:

-   **offsetX**: starting index for `x`.
-   **offsetY**: starting index for `y`.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the squared Euclidean distance between every other element in `x` starting from the second element with the last 3 elements in `y` in reverse order

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the distance between every other element in `x` starting from the second element with the last 3 elements in `y` in reverse order
```

```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If `N <= 0`, both functions return `NaN`.

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var dkmeansDistance = require( '@stdlib/ml/strided/dkmeans-distance' );

var opts = {
    'dtype': 'float64'
};
var x = discreteUniform( 10, 0, 100, opts );
console.log( x );

var y = discreteUniform( x.length, 0, 10, opts );
console.log( y );

var out = dkmeansDistance.ndarray( x.length, 'sqeuclidean', x, 1, 0, y, -1, y.length-1 );
console.log( out );
```

</section>

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/ml/strided/dkmeans_distance.h"
```

#### stdlib_strided_dkmeans_distance( N, metric, \*X, strideX, \*Y, strideY )

> Computes the distance between two double-precision floating-point strided arrays given a kmeans supported distance metric.

Reviewer: ```suggestion
Computes the distance between two double-precision floating-point strided arrays according to a specified distance metric.
```

```c
#include "stdlib/ml/base/kmeans/metrics.h"

const double x[] = { 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 };
const double y[] = { 2.0, 1.0, 2.0, 1.0, -2.0, 2.0, 3.0, 4.0 };

double v = stdlib_strided_dkmeans_distance( 8, STDLIB_ML_KMEANS_SQEUCLIDEAN, x, 1, y, 1 );
// returns 72.0
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.

> -   **metric**: `[in] enum STDLIB_ML_KMEANS_METRIC` distance metric. 

Reviewer: ```suggestion
-   **metric**: `[in] enum STDLIB_ML_KMEANS_METRIC` distance [metric][@stdlib/ml/base/kmeans/metrics]. 
```

-   **X**: `[in] double*` first input array.
-   **strideX**: `[in] CBLAS_INT` stride length of `X`.
-   **Y**: `[in] double*` second input array.
-   **strideY**: `[in] CBLAS_INT` stride length of `Y`.

```c
double stdlib_strided_dkmeans_distance( const CBLAS_INT N, const enum STDLIB_ML_KMEANS_METRIC, const double *X, const CBLAS_INT strideX, const double *Y, const CBLAS_INT strideY );
```

<!--lint disable maximum-heading-length-->

#### stdlib_strided_dkmeans_distance_ndarray( N, metric, \*X, strideX, offsetX, \*Y, strideY, offsetY )

<!--lint enable maximum-heading-length-->

> Computes the distance between two double-precision floating-point strided arrays given a kmeans supported distance metric using alternative indexing semantics.

Reviewer: ```suggestion
Computes the distance between two double-precision floating-point strided arrays according to a specified distance metric using alternative indexing semantics.
```

```c
#include "stdlib/ml/base/kmeans/metrics.h"

const double x[] = { 4.0, 2.0, -3.0, 5.0, -1.0 };
const double y[] = { 2.0, 6.0, -1.0, -4.0, 8.0 };

double v = stdlib_strided_dkmeans_distance_ndarray( 5, STDLIB_ML_KMEANS_SQEUCLIDEAN, x, -1, 4, y, -1, 4 );
// returns 186.0
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.

> -   **metric**: `[in] enum STDLIB_ML_KMEANS_METRIC` distance metric.

Reviewer: ```suggestion
-   **metric**: `[in] enum STDLIB_ML_KMEANS_METRIC` distance [metric][@stdlib/ml/base/kmeans/metrics].
```

    const double y[] = { 1.0, -2.0, 3.0, -4.0, 5.0, -6.0, 7.0, -8.0 };

    // Specify the number of elements:
    const int N = 8;

    // Specify strides:
    const int strideX = 1;
    const int strideY = -1;

    // Compute the distance between `x` and `y`:
    double d = stdlib_strided_dkmeans_distance( N, STDLIB_ML_KMEANS_SQEUCLIDEAN, x, strideX, y, strideY );

    // Print the result:
    printf( "Distance: %lf\n", d );

    // Compute the distance between `x` and `y` with offsets:
    d = stdlib_strided_dkmeans_distance_ndarray( N, STDLIB_ML_KMEANS_SQEUCLIDEAN, x, strideX, 0, y, strideY, N-1 );

    // Print the result:
    printf( "Distance: %lf\n", d );
}
```

</section>

<!-- /.examples -->

</section>

<!-- /.c -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[@stdlib/array/float64]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/float64

> [mdn-typed-array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray

Reviewer: ```suggestion
[mdn-typed-array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray

[@stdlib/ml/base/kmeans/metrics]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ml/base/kmeans/metrics
```
